packages/api/src/utils/w3up.js

```diff
@@ -0,0 +1,135 @@
+
+/**
+ * @param {object} env
+ * @param {string|undefined} [env.principal]
+ * @param {string|undefined} [env.proof]
+ */
+export async function getW3upClient({ principal, proof } = {}) {
+  const signer = principal ? ed25519.parse(principal) : await ed25519.generate()
+  const store = new StoreMemory()
+  const w3up = await W3UP.create({ principal: signer, store })
+  if (proof) {
+    await w3up.addSpace(await parseW3Proof(proof))
+  }
+  return { w3up, store, principal }
+}
+
+/**
+ * @param {string} proof - UCAN Delegation encoded as CID (multicodec=CAR, multihash=identity)
+ */
+async function parseW3Proof(proof) {
+  let cid
+  try {
+    cid = CID.parse(proof, base64)
+  } catch (/** @type {any} */ err) {
+    if (err?.message?.includes('Unexpected end of data')) {
+      console.error(
+        `Error: failed to read proof. The string has been truncated.`
+      )
+    }
+    throw err
+  }
+
+  if (cid.multihash.code !== identity.code) {
+    throw new Error('expected cid to have identity code')
+  }
+  const delegation = await readProofFromBytes(cid.multihash.digest)
+  return delegation
+}
+
+/**
+ * @param {Uint8Array} bytes Path to the proof file.
+ */
+export async function readProofFromBytes(bytes) {
+  const blocks = []
+  try {
+    const reader = await CarReader.fromBytes(bytes)
+    for await (const block of reader.blocks()) {
+      blocks.push(block)
+    }
+  } catch (/** @type {any} */ err) {
+    console.error(`Error: failed to parse proof: ${err.message}`)
+    throw err
+  }
+  try {
+    // @ts-expect-error Block types are slightly different but it works
+    return importDAG(blocks)
+  } catch (/** @type {any} */ err) {
+    console.error(`Error: failed to import proof: ${err.message}`)
+    throw err
+  }
+}
+
+/**
+ * @param {import('@ucanto/interface').Delegation} delegation - delegation to encode
+ */
+export async function encodeDelegationAsCid(delegation) {
+  const { writer, out } = CarWriter.create()
+  /** @type {Array<Uint8Array>} */
+  const carChunks = []
+  await Promise.all([
+    // write delegation blocks
+    (async () => {
+      for (const block of delegation.export()) {
+        // @ts-expect-error different Block types
+        await writer.put(block)
+      }
+      await writer.close()
+    })(),
+    // read out
+    (async () => {
+      for await (const chunk of out) {
+        carChunks.push(chunk)
+      }
+    })(),
+  ])
+  // now get car chunks
+  const car = new Blob(carChunks)
+  const bytes = new Uint8Array(await car.arrayBuffer())
+  const cid = CID.createV1(ucanto.CAR.code, identity.digest(bytes))
+  return cid
+}
+
+/**
+ * @param {object} options
+ * @param {string} options.url
+ * @param {string} options.principal
+ * @param {string} options.proof
+ */
+export async function createW3upClientFromConfig(options) {
+  const url = new URL(options.url)
+  const principal = ed25519.parse(options.principal)
+  const connection = connect({
+    id: { did: () => 'did:web:web3.storage' },
```

Comment:
Need service DID to come from env too. If we change `W3UP_URL`, then it won't match...e.g. for staging.

Reply:
done